Web logic against infantry

FS-21 opened the pull request:
- Infantry will be temporally paralyzed by warheads with `Webby=yes`.
- `Webby.Duration` specifies the duration, in frames, of the warhead's web effect.
- `Webby.DurationVariation` allows for random variance to the duration of the warhead's web effect.
- When `Webby.Anims` contains more than 1 animation then the new animation will be picked randomly.
- `Webby.Cap` works like in EMP logic developed by Ares.
  -  `Webby.Cap=-1` case: The target’s web counter is set to this absolute number of frames specified by `Web.Duration`, unless the target’s web counter is already greater than this.
  -  `Webby.Cap=0` case: Makes this web effect stackable, but uncapped.
  -  `Webby.Cap >0` case: Makes this web effect stackable, but maximum value capped to `Webby.Cap` value.
- Infantry can have custom values for the web logic.
- No damage is done by the weapons warhead.

In `rulesmd.ini`:
[SOMEWARHEAD]              ; Warhead
Webby=no                   ; boolean
Webby.Anims=               ; list of animations
Webby.Duration=0           ; integer, game frames
Webby.DurationVariation=0  ; integer
Webby.Cap=-1               ; integer

In `rulesmd.ini`:
[SOMETECHNO]               ; InfantryType
ImmuneToWeb=no             ; boolean
Webby.Anims=               ; list of animations
Webby.Duration=0           ; integer, game frames
Webby.DurationVariation=0  ; integer


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Infantry units can now be affected by web-based warheads with customizable effects, animations, and immunity options.
- **Documentation**
	- Updated CREDITS.md, New-or-Enhanced-Logics.md, and Whats-New.md to reflect the addition of web logic against infantry.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

FS-21 committed:
First commit

FS-21 committed:
Force AI to stop moving

And added documentation.

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/web-net

FS-21 committed:
More docs

FS-21 committed:
EditorConfig things

FS-21 commented on src/Ext/Techno/Body.h:
```diff
@@ -43,6 +43,12 @@ class TechnoExt
+		int	WebbyDurationCountDown;
+		CDTimerClass WebbyDurationTimer;
+		AnimClass* WebbyAnim;
+		AbstractClass* WebbyLastTarget;
```
I never touched this topic before, something like this?
![image](https://github.com/Phobos-developers/Phobos/assets/1371539/db5359a8-132c-4a31-ba4b-eed8e31dd71e)

FS-21 committed:
Added missing pointer invalidation

as suggested by Otamaa

MortonPL commented on docs/New-or-Enhanced-Logics.md:
```diff
@@ -1196,6 +1196,38 @@ In `rulesmd.ini`:
+- Infantry will be temporally paralyzed by warheads with `Webby=yes`.
+- `Webby.Duration` specifies the duration, in frames, of the warhead's web effect.
+- `Webby.DurationVariation` allows for random variance to the duration of the warhead's web effect.
+- When `Webby.Anims` contains more than 1 animation then the new animation will be picked randomly.
```
You should mention (perhaps as a {warning}?) that the anim should be set to loop forever, otherwise you'll have invisible infantry if the anim finishes before the webbed status ends.

FS-21 replied:
Good point. I didn't think in that detail

MortonPL commented on src/Ext/Techno/Hooks.cpp:
```diff
@@ -492,3 +494,116 @@ DEFINE_HOOK(0x51BAFB, InfantryClass_ChronoSparkleDelay, 0x5)
+	}
+
+	int duration = pTypeExt->Webby_Duration.Get() > 0 ? pTypeExt->Webby_Duration.Get() : pWarheadExt->Webby_Duration.Get();
+	int durationVariation = pTypeExt->Webby_DurationVariation.Get() > 0 ? pTypeExt->Webby_DurationVariation.Get() : pWarheadExt->Webby_DurationVariation.Get();
```
Might be a good idea to use Nullable<> here, but it's not important.

FS-21 replied:
I think in this case isn't necessary because then you'll have to ask before if .isset() and then if the value is over 0...

FS-21 replied:
ImmuneToWeb exists in the documentation.

In `rulesmd.ini`:
```ini
[SOMETECHNO]               ; InfantryType
ImmuneToWeb=no             ; boolean     <--------------------------
Webby.Anims=               ; list of animations
Webby.Duration=0           ; integer, game frames
Webby.DurationVariation=0  ; integer
```

MortonPL replied:
Yes, but there's no text description. Even if it's obvious, there should be at least a simple sentence like "Infantry with `ImmuneToWeb=no` are not affected by `Webby` warheads.".

MortonPL replied:
Also, use `false` instead of `no` in INI examples.

MortonPL replied:
And one thing I missed: it should be `[SOMEINFANTRY]`, not `[SOMETECHNO]`, even if it's internally a TechnoType trait only relevant to infantry.

FS-21 replied:
True!

MortonPL commented on src/Ext/Unit/Hooks.DisallowMoving.cpp:
```diff
@@ -81,3 +83,35 @@ DEFINE_HOOK(0x73891D, UnitClass_Active_Click_With_DisallowMoving, 0x6)
 	return pThis->Type->Speed == 0 ? 0x738927 : 0;
 }
+
+/*DEFINE_HOOK(0x51F663, InfantryClass_Mission_Move_DisallowMoving, 0x6)
+{
+	GET(InfantryClass*, pThis, ESI);
+
+	return pThis->ParalysisTimer.HasTimeLeft() > 0 ? 0x51F68F : 0;
+}*/
```
Whole code block comments either should be removed or described why they are here and why they are not used.

FS-21 replied:
I forgot to delete the latests tests -_-', obviously I have to delete it. Thanks for the feedback!

MortonPL reviewed (changes requested):
`ImmuneToWeb` should be described in the documentation. I think this PR is OK aside from a few review comments.

FS-21 committed:
Applied feedback

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
EditorConfig fixes

FS-21 committed:
Point to latest YRpp

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/web-net

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Fix bug

FS-21 committed:
Fix crash

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Removed 1 obsolete declaration

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/web-net

FS-21 committed:
Added ForceWeapon support

Added a new tag ForceWeapon.Webby=-1 that points the weapon that should use the unit against victims of the web net...

FS-21 commented:
Added a new tag ForceWeapon.Webby=-1 that points the weapon that should use the unit against victims of the web net...

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

TaranDahl commented on src/Ext/Unit/Hooks.DisallowMoving.cpp:
```diff
@@ -81,3 +83,27 @@ DEFINE_HOOK(0x73891D, UnitClass_Active_Click_With_DisallowMoving, 0x6)
+		return 0x51B1D7;
+
+	auto pExt = TechnoExt::ExtMap.Find(pThis);
+	if (!pExt)
```
same

TaranDahl commented on src/Ext/Techno/Hooks.cpp:
```diff
@@ -654,3 +654,116 @@ DEFINE_HOOK(0x465D40, BuildingClass_Is1x1AndUndeployable_BuildingMassSelectable,
+		return 0;
+
+	if (pExt->WebbyDurationCountDown > 0)
+		return SkipDrawCode;
```
If I understand correctly, should this logic be similar to the webs thrown by the Wolf Rider and the Spider in Warcraft 3? 
Then should the image of the web be attached to the unit like a parachute, rather than directly replacing the unit's image?

FS-21 replied:
https://youtube.com/clip/Ugkxp3Txb_BCE0Re40RY83sqlRm9gWBUZRrk?si=9PfCNvwtFncDZnxR

TaranDahl replied:
Got it.

TaranDahl commented on src/Ext/Techno/Hooks.cpp:
```diff
@@ -654,3 +654,116 @@ DEFINE_HOOK(0x465D40, BuildingClass_Is1x1AndUndeployable_BuildingMassSelectable,
+		return 0;
+
+	auto pExt = TechnoExt::ExtMap.Find(pTechno);
+	if (!pExt)
```
same

TaranDahl commented on src/Ext/Techno/Hooks.cpp:
```diff
@@ -654,3 +654,116 @@ DEFINE_HOOK(0x465D40, BuildingClass_Is1x1AndUndeployable_BuildingMassSelectable,
+		return 0;
+
+	auto const pTypeExt = TechnoTypeExt::ExtMap.Find(pThis->GetTechnoType());
+	if (!pTypeExt)
```
same

TaranDahl commented on docs/New-or-Enhanced-Logics.md:
```diff
@@ -1867,6 +1867,44 @@ In `rulesmd.ini`:
+  -  `Webby.Cap >0` case: Makes this web effect stackable, but maximum value capped to `Webby.Cap` value.
+- Infantry can have custom values for the web logic.
+- Infantry with `ImmuneToWeb=true` are not affected by `Webby=true` warheads.
+- No damage is done by the weapons warhead.
```
Any reason for skipping the damage?

TaranDahl replied:
I think it's best for the two systems to work independently of each other unless there is a necessary interaction requirement. This logic shouldn't have an impact on damage, because there is no necessary interaction requirement here.

FS-21 replied:
with the latest changes it should work as you mentioned

TaranDahl commented on src/Ext/Techno/Hooks.cpp:
```diff
@@ -654,3 +654,116 @@ DEFINE_HOOK(0x465D40, BuildingClass_Is1x1AndUndeployable_BuildingMassSelectable,
+	return 0;
+}
+
+DEFINE_HOOK(0x518016, InfantryClass_TakeDamage_Webby, 0x7)
```
In Phobos (as well as Ares), things like AE or EMP are delivered on `BulletClass_Detonate`. Aligning with them could be better.
It could be much better if you make this logic a part of Phobos AE.

FS-21 replied:
done! (not the AE thingy)

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Compilation fix

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 commented on docs/New-or-Enhanced-Logics.md:
```diff
@@ -1867,6 +1867,44 @@ In `rulesmd.ini`:
+Webby.DurationVariation=0  ; integer
+Webby.Cap=-1               ; integer
+
+[SOMEINFANTRY]             ; InfantryType
```
Because you already have similar features (EMP) for vehicles but not the Firestorm's Webby logic.

TaranDahl replied:
I watched the video you sent. Maybe there's really no need to make it take effect for other types.

FS-21 commented on src/Ext/Techno/Hooks.Firing.cpp:
```diff
@@ -102,6 +102,14 @@ DEFINE_HOOK(0x6F3428, TechnoClass_WhatWeaponShouldIUse_ForceWeapon, 0x6)
+		{
+			if (const auto pTargetExt = TechnoExt::ExtMap.Find(pTarget))
+			{
+				if (pTargetExt->WebbyAnim)
```
Sorry TaranDahl but "same" what?

TaranDahl replied:
Checking the timer makes more sense.

FS-21 committed:
Applied some feedback

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/web-net

FS-21 committed:
Added Map Event when a tagget soldier is under Webby logic

607,0,0

TaranDahl commented on docs/New-or-Enhanced-Logics.md:
```diff
@@ -1867,6 +1867,44 @@ In `rulesmd.ini`:
+[SOMEINFANTRY]             ; InfantryType
+ImmuneToWeb=false          ; boolean
+Webby.Anims=               ; list of animations
+Webby.Duration=0           ; integer, game frames
```
If I understand correctly, it's some kind of debuff resistance. It should be multiplier instead of countdown.

FS-21 replied:
applied feedback for infantry.

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/web-net

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
YRpp update

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 committed:
Fixed a crash

FS-21 committed:
Merge branch 'develop' into feature/web-net

FS-21 commented on src/Ext/Techno/Body.Update.cpp:
```diff
@@ -43,6 +43,9 @@ void TechnoExt::ExtData::OnEarlyUpdate()
 	this->DepletedAmmoActions();
 	this->UpdateAttachEffects();
+
+	if (this->WebbyAnim)
```
done
